src/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -1433,6 +1433,79 @@
                    (when (and update-latest-report? (not= type "plan"))
                      (update-latest-report! certname report-id producer_timestamp)))))))))
 
+(defn spawn-query-killer
+  "Spawn a thread which will loop and wait for the gc-pid to get blocked
+   waiting on an AccessExclusiveLock. Once this thread detects that GC is
+   blocked it will cancel all running queries from the pdb user against the
+   pdb database which have granted locks with the exception of those queries
+   from the gc-pid or the killer's pid. This should clear the way for GC to
+   grab the lock it's requesting in the main GC thread. This loop repeats
+   until the killer thread is interrupted or receives confirmation from the
+   GC thread via the gc-finished? atom that the GC thread has dropped the
+   partition."
+  [db gc-pid gc-finished?]
+  (let [{:keys [current_database current_user]}
+        (first (jdbc/query-to-vec "select current_database(), current_user;"))
+        kill-blocking-qs (format
+                          "select pg_cancel_backend(l.pid)
+                            from pg_locks l inner join pg_stat_activity a on l.pid = a.pid
+                            where (a.datname = '%s' AND a.usename = '%s')
+                             and not (l.pid = pg_backend_pid() OR l.pid = %s)
+                             and l.granted = true
+                             and exists (select 1
+                                          from pg_locks
+                                          where pid = %s and mode = 'AccessExclusiveLock'
+                                           and granted = false)
+                            group by l.pid;"
+                          current_database current_user gc-pid gc-pid)]
+    (Thread.
+     #(with-noisy-failure
+        (try
+          ;; you can't use *db* value from the gc thread because the connection
+          ;; it has may already be blocked attempting to drop the partition.
+          ;; Binding *db* to the passed in db var will cause the killer thread
+          ;; to get a new Hikari connection
+          (binding [jdbc/*db* db]
+            ;; Q: is there any possible way that all queries could be in a blocked
+            ;; state where we're unable to get another connection from the pool
+            ;; to issue the kill command?
+            (while (not @gc-finished?)
+              (jdbc/do-commands kill-blocking-qs)
+              (Thread/sleep 1000)))
+          ;; Q: do we need to worry about other exceptions here as well?
```

Comment:
We could wrap it all with `with-nonfatal-exceptions-suppressed` but I don't know if that will have any difference either way the Thread is going to die and with-noisy-failure will ensure it is logged.

Reply:
👍 I'll double check what `with-nonfatal-exceptions-suppressed` is doing and see if we want it here.